descartes_light/core/include/descartes_light/types.h

```diff
@@ -3,28 +3,188 @@
 #include <descartes_light/descartes_macros.h>
 DESCARTES_IGNORE_WARNINGS_PUSH
 #include <vector>
+#include <memory>
 #include <Eigen/Geometry>
 DESCARTES_IGNORE_WARNINGS_POP
 
 namespace descartes_light
 {
-/** @brief Definition of a state */
 template <typename FloatType>
-using State = Eigen::Matrix<FloatType, Eigen::Dynamic, 1>;
+using Array = Eigen::Array<FloatType, Eigen::Dynamic, 1>;
 
+/** @brief Definition of a state */
 template <typename FloatType>
-using Array = Eigen::Array<FloatType, Eigen::Dynamic, 1>;
+class State
+{
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  using Ptr = std::shared_ptr<State>;
+  using ConstPtr = std::shared_ptr<const State>;
+
+  State() = default;
+  virtual ~State() = default;
+  State(const State&) = delete;
+  State& operator=(const State&) = delete;
+  State(State&&) noexcept = delete;
+  State& operator=(State&&) noexcept = delete;
+
+  ///////////////////////////
+  // Joint State Container //
+  ///////////////////////////
+  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> values;
+
+  ////////////////////////
+  // Eigen Constructors //
+  ////////////////////////
+
+  // This constructor allows you to construct MyVectorType from Eigen expressions
+  template <typename OtherDerived>
+  State(const Eigen::MatrixBase<OtherDerived>& other) : values(other)
+  {
+  }
+
+  State(std::initializer_list<double> l)
+  {
+    values.resize(static_cast<Eigen::Index>(l.size()));
+    Eigen::Index i = 0;
+    for (auto& v : l)
+      values(i++) = v;
+  }
+
+  ///////////////////
+  // Eigen Methods //
+  ///////////////////
+
+  /** @returns true if two are approximate */
+  inline Eigen::Index size() const { return values.size(); }
+  /** @returns norm of vector */
+  inline double norm() const { return values.norm(); }
+  /** @returns true if two are approximate */
+  inline bool isApprox(const Eigen::Matrix<FloatType, Eigen::Dynamic, 1>& other,
+                       FloatType prec = FloatType(1e-12)) const
```

Comment:
The default precision here seems arbitrary. We should use the `numeric_limits` or Eigen::VectorX default instead
```suggestion
  inline bool isApprox(const Eigen::Matrix<FloatType, Eigen::Dynamic, 1>& other,
                       FloatType prec = std::numeric_limits<FloatType>::epsilon()) const
```

Reply:
This is what Eigen uses so I kept it the same. This would change the behavior compared to Eigen isApprox.